src/ir/subtype-exprs.h

```diff
@@ -265,8 +265,18 @@ struct SubtypingDiscoverer : public OverriddenVisitor<SubType> {
   }
   void visitRethrow(Rethrow* curr) {}
   void visitThrowRef(ThrowRef* curr) {}
-  void visitTupleMake(TupleMake* curr) {}
-  void visitTupleExtract(TupleExtract* curr) {}
+  void visitTupleMake(TupleMake* curr) {
+    if (curr->type != Type::unreachable) {
+      for (Index i = 0, size = curr->type.size(); i < size; ++i) {
+        self()->noteSubtype(curr->operands[i], curr->type[i]);
+      }
+    }
+  }
+  void visitTupleExtract(TupleExtract* curr) {
+    if (curr->type != Type::unreachable) {
+      self()->noteSubtype(curr->tuple->type[curr->index], curr->type);
```

Comment:
Is this last one needed? This is really an equality here, not a subtyping, and more of a validation issue. Put another way, it is the same as `visitStructGet`, which is intentionally empty.

Reply:
The problem is that we allow operands to be refined without refining their parents in valid IR, so utilities that work on valid IR have to be conservative and assume that might have happened. In other words we should probably update StructGet here, too.

Alternatively we could disallow stale types entirely in valid IR, since we try to avoid them in practice anyway. That would mean that all the passes that fail to refinalize their global initializers would be recharacterized as buggy, so the bug here would be on TypeSSA instead of Unsubtyping.

Reply:
We do want to disallow stale types in general, I believe, just checking that is expensive so we only do it when pass-debug asserts are on. So, yes, I agree this is a TypeSSA bug - it should refinalize module-level code.

Given that, yeah, I guess both tuple changes here are not actually necessary, as they both (I thought only one, but re-reading now I see it is both) are just equalities, not actual constraints on subtyping.

Reply:
It is possible we currently fail to check pass-debug staleness on module-level code, btw - that might explain why this was missed actually. Worth checking that.

Reply:
No, it's because our staleness check explicitly allows refinement of operands. I'll try changing it to require type equality and see what breaks.

Reply:
It turns out that this breaks 104 tests. Will dig in after lunch to characterize why.